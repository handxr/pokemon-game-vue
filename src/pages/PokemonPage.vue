<template>
  <div class="container" v-if="isLoading">
    <p>Loading...</p>
  </div>
  <div class="container" v-else>
    <h1>¿Cuál es este Pokemon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonOptions" @selectPokemon="checkAnswer" />
    <p v-if="showPokemon">{{ message }}</p>
    <button v-if="showPokemon" @click="newGame">Nuevo juego</button>
  </div>
</template>

<script>
import PokemonOptions from '../components/PokemonOptions.vue'
import PokemonPicture from '../components/PokemonPicture.vue'
import { getPokemonOptions } from '../helpers/getPokemonOptions'

export default {
  components: { PokemonPicture, PokemonOptions },

  data() {
    return {
      pokemonOptions: [],
      pokemon: null,
      isLoading: false,
      showPokemon: false,
      isCorrectAnswer: false,
      message: ''
    }
  },

  methods: {
    async getPokemonOptions() {
      try {
        this.isLoading = true
        const pokemonOptions = await getPokemonOptions()
        this.pokemonOptions = pokemonOptions
        const randomIndex = Math.floor(Math.random() * 4)
        this.pokemon = pokemonOptions[randomIndex]
      } catch (error) {
        console.log(error)
      } finally {
        this.isLoading = false
      }
    },
    checkAnswer(pokemon) {
      if (pokemon.id === this.pokemon.id) {
        this.isCorrectAnswer = true
        this.showPokemon = true
        this.message = 'Correcto!'
      } else {
        this.isCorrectAnswer = false
        this.showPokemon = true
        this.message = `Incorrecto! El pokemon correcto es ${this.pokemon.name}`
      }
    },
    newGame() {
      this.showPokemon = false
      this.getPokemonOptions()
    }
  },

  created() {
    this.getPokemonOptions()
  }
}
</script>

<style scoped>
.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: start;
  height: 100vh;
}
</style>
