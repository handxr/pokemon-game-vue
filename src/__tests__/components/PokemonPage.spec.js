import PokemonPage from '@/pages/PokemonPage.vue'
import { shallowMount, mount } from '@vue/test-utils'
import { it, describe, expect, beforeEach, vitest } from 'vitest'
import { pokemons } from '../__mocks__/pokemons'

describe.only('PokemonPage component', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallowMount(PokemonPage)
  })
  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should call the getPokemonOptions function when the component is mounted', () => {
    const getPokemonOptions = vitest.spyOn(PokemonPage.methods, 'getPokemonOptions')
    mount(PokemonPage)
    expect(getPokemonOptions).toHaveBeenCalled()
  })
  it('should match the snapshot when the component is mounted', () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonOptions: pokemons,
          pokemon: pokemons[0],
          isLoading: false,
          showPokemon: false,
          isCorrectAnswer: false,
          message: ''
        }
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
