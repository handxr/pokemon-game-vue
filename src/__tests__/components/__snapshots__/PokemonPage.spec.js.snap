// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonPage component > should match the snapshot 1`] = `
"<div data-v-2108564c=\\"\\" class=\\"container\\">
  <p data-v-2108564c=\\"\\">Loading...</p>
</div>"
`;

exports[`PokemonPage component > should match the snapshot when the component is mounted 1`] = `
"<div data-v-2108564c=\\"\\" class=\\"container\\">
  <p data-v-2108564c=\\"\\">Loading...</p>
</div>"
`;
